background/main.ts

```diff
@@ -508,27 +508,82 @@ export default class Main extends BaseService<never> {
     // Start up the redux store and set it up for proxying.
     this.store = initializeStore(savedReduxState, this)
 
+    /**
+     * Tracks pending updates to the redux store. This is used to delay responding
+     * to dispatched thunks until the store has been updated in the UI. This is
+     * necessary to prevent race conditions where the UI expects the store to be
+     * updated before the thunk has finished dispatching.
+     */
+    let storeUpdateLock: Promise<void> = Promise.resolve()
+    const pendingUpdates: (() => void)[] = []
+
+    const queueUpdate = debounce(
+      (lastState, newState, updateFn) => {
+        const clearPendingUpdates = () => {
+          pendingUpdates.forEach((resolve) => resolve())
+
+          // Clear all currently tracked updates
+          while (pendingUpdates.length) {
+            pendingUpdates.pop()
+          }
```

Comment:
If I'm not mistaken and the below remark about not needing an array doesn't apply, I think this can be rewritten as:

```suggestion
          pendingUpdates
            .splice(0)
            .forEach((resolve) => resolve())
```